imap_processing/utils.py

```diff
@@ -228,29 +228,77 @@ def create_dataset(
     return dataset
 
 
-def update_epoch_to_datetime(dataset: xr.Dataset) -> xr.Dataset:
+def packet_file_to_datasets(
+    packet_file: Union[str, Path],
+    xtce_packet_definition: Union[str, Path],
+    use_derived_value: bool = True,
+) -> dict[int, xr.Dataset]:
     """
-    Update epoch in dataset to datetime object.
+    Convert a packet file to xarray datasets.
+
+    The packet file can contain multiple apids and these will be separated
+    into distinct datasets, one per apid. The datasets will contain the
+    ``derived_value``s of the data fields, and the ``raw_value``s if no
+    ``derived_value`` is available. If there are conversions in the XTCE
+    packet definition, the ``derived_value`` will be the converted value.
+    The dimension of the dataset will be the time field in J2000 nanoseconds.
 
     Parameters
     ----------
-    dataset : xr.Dataset
-        Dataset to update.
+    packet_file : str
+        Path to data packet path with filename.
+    xtce_packet_definition : str
+        Path to XTCE file with filename.
+    use_derived_value : bool, default True
+        Whether or not to use the derived value from the XTCE definition.
 
     Returns
     -------
-    dataset : xr.Dataset
-        Dataset with updated epoch dimension from int to datetime object.
+    datasets : dict
+        Mapping from apid to xarray dataset, one dataset per apid.
     """
-    # convert epoch to datetime
-    epoch_converted_time = met_to_j2000ns(dataset["epoch"])
-    # add attrs back to epoch
-    epoch = xr.DataArray(
-        epoch_converted_time,
-        name="epoch",
-        dims=["epoch"],
-        attrs=ConstantCoordinates.EPOCH,
-    )
-    dataset = dataset.assign_coords(epoch=epoch)
+    # Set up containers to store our data
+    # We are getting a packet file that may contain multiple apids
+    # Each apid has consistent data fields, so we want to create a
+    # dataset per apid.
+    # {apid1: dataset1, apid2: dataset2, ...}
+    data_dict: dict[int, dict] = dict()
+
+    # Set up the parser from the input packet definition
+    packet_definition = xtcedef.XtcePacketDefinition(xtce_packet_definition)
+    packet_parser = parser.PacketParser(packet_definition)
+
+    with open(packet_file, "rb") as binary_data:
+        packet_generator = packet_parser.generator(binary_data)
+        for packet in packet_generator:
+            apid = packet.header["PKT_APID"].raw_value
+            if apid not in data_dict:
+                # This is the first packet for this APID
+                data_dict[apid] = collections.defaultdict(list)
+
+            # TODO: Do we want to give an option to remove the header content?
+            packet_content = packet.data | packet.header
```

Comment:
I do think it will be helpful to add that option in this function. Removing packet header content will be same across all packet and makes sense to do that here. As for removing data following header, it makes sense to do that in the individual processing code as you mentioned.

Reply:
I am starting to be more of the mindset that we don't want to be adding options just to have them, is there a specific instrument/use-case you know of where we want both cases. So for now, I think keeping all packet content seems reasonable and then downstream people can remove the header if they want. But it seems like more people wanted the header than not?

Reply:
I have not heard strong preference of wanting to remove packet header. I think it came up in past discussion and brought it up just in case. That makes sense and works for me.